File: src/git.ts
import { ExecOptions, exec } from '@actions/exec';
import { log } from './logger';

interface ExecResult {
  stdout: string;
  stderr: string;
  code: number | null;
}

const capture = async (cmd: string, args: string[], options: ExecOptions = {}): Promise<ExecResult> => {
  const res: ExecResult = {
    stdout: '',
    stderr: '',
    code: null,
  };

  try {
    const code = await exec(cmd, args, {
      ...options,
      listeners: {
        stdout(data) {
          res.stdout += data.toString();
        },
        stderr(data) {
          res.stderr += data.toString();
        },
      },
    });
    res.code = code;
    return res;
  } catch (err) {
    const msg = `Command '${cmd}' failed with args '${args.join(' ')}': ${res.stderr}: ${err}`;
    log.debug(`@actions/exec.exec() threw an error: ${msg}`);
    throw new Error(msg);
  }
};

export const findTargetHash = async (baseSha: string, headSha: string): Promise<string> => {
  log.info(`base sha is ${baseSha}, head sha is ${headSha}`);

  await capture('git', ['config', 'remote.origin.fetch', '+refs/heads/*:refs/remotes/origin/*']);

  await capture('git', ['fetch', '--all']);

  const args = ['merge-base', '-a', `${baseSha}`, `${headSha}`];

  const res = await capture('git', args);

  if (res.code !== 0) {
    throw new Error(`Command 'git ${args.join(' ')}' failed: ${JSON.stringify(res)}`);
  }
  const targetHash = res.stdout;
  return targetHash;
};

export const configureName = async (name: string): Promise<void> => {
  await capture('git', ['config', '--global', 'user.name', name]);
};

export const configureEmail = async (email: string): Promise<void> => {
  await capture('git', ['config', '--global', 'user.email', email]);
};

export const clone = async (input: { repo: string; dist: string }, options: ExecOptions = {}): Promise<ExecResult> => {
  return capture('git', ['clone', input.repo, input.dist], options);
};

export const fetchOrigin = async (input: { branch: string }, options: ExecOptions = {}): Promise<ExecResult> => {
  return capture('git', ['fetch', '-u', 'origin', `${input.branch}:${input.branch}`], options);
};

export const hasBranch = async (branch: string, options: ExecOptions = {}): Promise<boolean> => {
  const res = await capture('git', ['branch', '--list', branch], options);
  return res.stdout.trim() !== '';
};

export const checkout = async (branch: string, orphan: boolean, options: ExecOptions = {}): Promise<ExecResult> => {
  const args = orphan ? ['checkout', '--orphan', branch] : ['checkout', branch];
  return capture('git', args, options);
};

export const add = async (options: ExecOptions = {}): Promise<ExecResult> => {
  return capture('git', ['add', '-A', '.'], options);
};

export const commit = async (message: string, options: ExecOptions = {}): Promise<ExecResult> => {
  return capture('git', ['commit', '-m', message], options);
};

export const push = async (branch: string, options: ExecOptions = {}): Promise<ExecResult> => {
  return capture('git', ['push', 'origin', branch], options);
};

export const rebase = async (branch: string, options: ExecOptions = {}): Promise<ExecResult> => {
  return capture('git', ['rebase', `origin/${branch}`], options);
};

File: src/git.rs
//! `git` CLI ラッパ。サブプロセス実行で済ませる (gix への移行は将来)。

use std::path::Path;
use std::process::{Command, Stdio};

use anyhow::{anyhow, Context, Result};

#[derive(Debug)]
pub struct ExecResult {
    pub stdout: String,
    pub stderr: String,
    pub code: i32,
}

pub fn capture<I, S, P>(args: I, cwd: Option<&P>) -> Result<ExecResult>
where
    I: IntoIterator<Item = S>,
    S: AsRef<std::ffi::OsStr>,
    P: AsRef<Path>,
{
    let mut cmd = Command::new("git");
    cmd.args(args);
    if let Some(d) = cwd {
        cmd.current_dir(d);
    }
    let out = cmd
        .stdout(Stdio::piped())
        .stderr(Stdio::piped())
        .output()
        .context("failed to spawn `git`")?;
    Ok(ExecResult {
        stdout: String::from_utf8_lossy(&out.stdout).into_owned(),
        stderr: String::from_utf8_lossy(&out.stderr).into_owned(),
        code: out.status.code().unwrap_or(-1),
    })
}

fn ensure_ok(r: &ExecResult, hint: &str) -> Result<()> {
    if r.code != 0 {
        return Err(anyhow!(
            "git {hint} failed (code={}): stderr={}",
            r.code,
            r.stderr.trim()
        ));
    }
    Ok(())
}

/// PR の base..head の merge-base を求める。
pub fn find_target_hash(base_sha: &str, head_sha: &str) -> Result<String> {
    tracing::info!(base = base_sha, head = head_sha, "find merge-base");
    let _ = capture::<_, _, &Path>(
        [
            "config",
            "remote.origin.fetch",
            "+refs/heads/*:refs/remotes/origin/*",
        ],
        None,
    )?;
    let _ = capture::<_, _, &Path>(["fetch", "--all"], None)?;
    let r = capture::<_, _, &Path>(["merge-base", "-a", base_sha, head_sha], None)?;
    ensure_ok(&r, "merge-base")?;
    Ok(r.stdout.trim().to_string())
}

pub fn configure_name(name: &str) -> Result<()> {
    let r = capture::<_, _, &Path>(["config", "--global", "user.name", name], None)?;
    ensure_ok(&r, "config user.name")
}

pub fn configure_email(email: &str) -> Result<()> {
    let r = capture::<_, _, &Path>(["config", "--global", "user.email", email], None)?;
    ensure_ok(&r, "config user.email")
}

pub fn clone(repo_url: &str, dest: &Path) -> Result<()> {
    let r = capture::<_, _, &Path>(["clone", repo_url, &dest.to_string_lossy()], None)?;
    ensure_ok(&r, "clone")
}

pub fn fetch_origin(branch: &str, cwd: &Path) -> Result<ExecResult> {
    capture(
        ["fetch", "-u", "origin", &format!("{branch}:{branch}")],
        Some(&cwd),
    )
}

pub fn has_branch(branch: &str, cwd: &Path) -> Result<bool> {
    let r = capture(["branch", "--list", branch], Some(&cwd))?;
    Ok(!r.stdout.trim().is_empty())
}

pub fn checkout(branch: &str, orphan: bool, cwd: &Path) -> Result<()> {
    let r = if orphan {
        capture(["checkout", "--orphan", branch], Some(&cwd))?
    } else {
        capture(["checkout", branch], Some(&cwd))?
    };
    ensure_ok(&r, "checkout")
}

pub fn add_all(cwd: &Path) -> Result<()> {
    let r = capture(["add", "-A", "."], Some(&cwd))?;
    ensure_ok(&r, "add")
}

pub fn commit(message: &str, cwd: &Path) -> Result<()> {
    let r = capture(["commit", "-m", message], Some(&cwd))?;
    ensure_ok(&r, "commit")
}

pub fn push(branch: &str, cwd: &Path) -> Result<ExecResult> {
    capture(["push", "origin", branch], Some(&cwd))
}

pub fn rebase(branch: &str, cwd: &Path) -> Result<ExecResult> {
    capture(["rebase", &format!("origin/{branch}")], Some(&cwd))
}
